util/tls/tls_socket.cc

```diff
@@ -331,22 +331,254 @@ io::Result<TlsSocket::PushResult> TlsSocket::PushToEngine(const iovec* ptr, uint
   return res;
 }
 
-// TODO: to implement async functionality.
+void TlsSocket::AsyncWriteReq::CompleteAsyncReq(io::Result<size_t> result) {
+  auto current = std::exchange(owner->async_write_req_, std::nullopt);
+  current->caller_completion_cb(result);
+}
+
+void TlsSocket::AsyncReadReq::CompleteAsyncReq(io::Result<size_t> result) {
+  auto current = std::exchange(owner->async_read_req_, std::nullopt);
+  current->caller_completion_cb(result);
+}
+
+void TlsSocket::AsyncReqBase::HandleOpAsync(int op_val) {
+  switch (op_val) {
+    case Engine::EOF_STREAM:
+      VLOG(1) << "EOF_STREAM received " << owner->next_sock_->native_handle();
+      CompleteAsyncReq(make_unexpected(make_error_code(errc::connection_aborted)));
+      break;
+    case Engine::NEED_READ_AND_MAYBE_WRITE:
+      MaybeSendOutputAsyncWithRead();
+      break;
+    case Engine::NEED_WRITE:
+      MaybeSendOutputAsync();
+      break;
+    default:
+      LOG(DFATAL) << "Unsupported " << op_val;
+  }
+}
+
+void TlsSocket::AsyncWriteReq::Run() {
```

Comment:
why do we have `while (true)` here but we return in most cases?
it's hard to understand this logic - at first I thought Run is fiber-blocking.

Reply:
what I am trying to understand - what is the path to the second loop?

Reply:
There are no fiber blocks, everything is async and yes we need that. An example would be:

1. We push to the engine
2. We call HandleOpAsync because last_push.engine_opcode < 0 and we submit an async operation to the underline socket. Its completion handler will call `Run()` again.
3. The fiber continues doing its work while the underline async operation is in flight
4. The async operation completes, invokes the completion handler which calls `Run()` again. Our state now is `MaybeSendOutputAsyncTag` so we reach that branch but the condition `last_push.written() > 0` is not true. Now we loop, push to the engine again and proceed.